### ISSUE TITLE
Variables may live longer than they suppose to on Eager Execution

### ISSUE THREAD

**Issue by julesbayet** [tensorflow/tensorflow#19671]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: 1.7.0
- **Python version**: 3.6

### Describe the problem

Using tensorflow EagerExecution. According to the documentation :

> During eager execution the lifetime of state objects is determined by the lifetime of their corresponding Python object.

I have some trouble with how tensorflow handle memory. I would like to remove tensors from my memory after each iteration on this toy example. The results are shown in the chart.

I have tried with Variables and with simple tensors. tf.assign doesn't do the job. More and more memory is used. It might be normal in order to be able to compute the gradient. But, if I apply some dummy optimizer at the end of each iteration, the memory isn't not released (more precisely, it happens sometimes but the global trend is that the memory use is growing).

I haven't found any API to deal with that yet.

![stackoverflow](https://user-images.githubusercontent.com/38664274/40790405-c17a6b2c-64ec-11e8-8a35-361885acd44a.PNG)

### Source code / logs

```
import tensorflow as tf
import tensorflow.contrib.eager as tfe
import numpy as np
import time as ti


tf.enable_eager_execution()

for i in range(150):
    all_subject=tfe.Variable(np.random.rand(200, 500), dtype=tf.float32)
    tf.assign(all_subject, np.random.rand(200,500) )
    ti.sleep(1.0)
    del all_subject
    ti.sleep(0.5)
```

**Comment by allenlavoie** [comment]
I was looking at a related issue last week and was going to try fixing this one soon. Sorry you ran into it first!

At least one reason variable creation in a loop leaks is caching of the resource creation kernels (cached keyed by attributes, which include a unique string). Shouldn't be too hard to fix, either with a no-attribute resource creation op or by ejecting some from the cache.

**Comment by julesbayet** [comment]
Thanks for taking care of the problem. Do you have any news on the topic ?

**Comment by allenlavoie** [comment]
If you want a workaround in the meantime, `tf.set_random_seed(1)` will clear the kernel cache.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lukostaz** [comment]
Hello @allenlavoie. Was wondering if you had time to look at this by any chance.

**Comment by allenlavoie** [comment]
It's still on my list. Happy to explain one approach to fixing it if you (or someone else) wants to pick it up in the meantime (essentially we need to add an "anonymous" resource creation op for variables which does not take a shared name or container name as an attribute but generates these itself, like [AnonymousIteratorHandleOp](https://github.com/tensorflow/tensorflow/commit/70674b950ab48f913ed1c99e48c4162287595d46); that kernel will get cached just once across all variable creation).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tuananhle7** [comment]
Are there any news on this issue? I cannot use the `tf.set_random_seed(1)` work around as my training depends on randomness.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 49 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 64 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 79 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @allenlavoie: It has been 94 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by feihugis** [comment]
@allenlavoie Is this issue fixed? If not, I would like to work on it.

**Comment by allenlavoie** [comment]
It has not been fixed, although I believe @jaingaurav had started on a fix a week or so ago. So you may want to coordinate with him.

**Comment by feihugis** [comment]
@allenlavoie Thanks for your quick reply! @jaingaurav Please feel free to let me know if there is anything I can do.

**Comment by jaingaurav** [comment]
Thanks for the help @feihugis. I should have the fix up in the next few days, justing adding tests for it at the moment.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jaingaurav: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jaingaurav** [comment]
Just an update for those interested in this issue. Still working on this issue, however upon further investigation, there are an number of additional ops that are demonstrating the same issue. As a result, I'm trying to come up with a more comprehensive fix for all affected ops.

**Comment by feihugis** [comment]
@jaingaurav Thanks for your updates! I'm very interested in this issue. Could you share more details here? Is it related to https://github.com/tensorflow/community/blob/master/rfcs/20180817-variables-20.md?

**Comment by vtolani95** [comment]
@jaingaurav Thanks for your work thus far. I've run into this same issue myself recently on a project. 

Is the fix still in progress?

**Comment by jaingaurav** [comment]
Hi, yes this issue was on the back burner for a bit but I'm currently working through an updated change internally.

@feihugis: There is some relation to the doc you referenced. However, the issue is more specific to eager execution. At https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/resource_variable_ops.py#L409 we generate unique shared names for eager variables. As a result, the kernel cache expands as more and more eager variables are created since we cache based on the attributes. As @allenlavoie mentioned, `tf.set_random_seed(1)` will clear the kernel cache.

**Comment by feihugis** [comment]
@jaingaurav Thanks for your explanation. The following issues may also be related to the kernel cache: #23882, #23407, and #24084.

**Comment by jaingaurav** [comment]
@feihugis, @julesbayet: Fix has been landed. Thanks for your patience with this fix. Please let me know if you are still seeing the issue after my change.

**Comment by araujoalexandre** [comment]
Hi, has the update been done on v1.13?
Thank you,

**Comment by jaingaurav** [comment]
Hi @araujoalexandre, the fix did not make it in the v1.13 release. It should be in the next release. You can also try the nightlies to verify that it has indeed been fixed.

**Comment by araujoalexandre** [comment]
Thank you @jaingaurav, I have access to the v1.12 and v1.13, is it possible to fix this issue an other way ? I tried tf.set_random_seed(1) but is not working

**Comment by jaingaurav** [comment]
@araujoalexandre: `tf.set_random_seed(1)` should have addressed this issue. Would you be able to share code? Perhaps there is something else that is causing a memory leak for you.

**Comment by araujoalexandre** [comment]
@jaingaurav I am try to do gradient descent in a eager py_func, in this function, I need to make several calls to the model. I think every time I make a call a copy of the variables is made.
You can find a run code here -> https://github.com/araujoalexandre/tf_mem_leak/blob/master/memory_leak.py

You can run it with :
mkdir train_folder; python3 memory_leak.py --train_dir ./train_folder/

I am using tensorflow 1.12
Thank you,

**Comment by XuesongYang** [comment]
Just to need to be confirm, `tf.set_random_seed(1)` should be applied within the for-loop, right?

**Comment by jaingaurav** [comment]
@XuesongYang: setting `tf.set_random_seed(1)` can be set at anytime to free the memory. I would use the workaround with caution though as it might change the behavior of your program. An alternative would be to use the nightly pip package or wait till the 1.14 release.

**Comment by ChenYang-ChenYang** [comment]
@jaingaurav I'm using tensorflow 2.0.0b1 keras with eager mode. I can't find tf.set_random_seed(). Where is it in version 2.0? Thanks.

**Comment by ChenYang-ChenYang** [comment]
> > @jaingaurav I'm using tensorflow 2.0.0b1 keras with eager mode on GPU and got similar issue that the memory keep growing till OOM. I can't find tf.set_random_seed() so I tried to use tf.random.set_seed(1) but it didn't work. Any suggestion? Thanks.
> 
> [`tf.random.set_seed`](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/random/set_seed)

@velocirabbit Thanks for the reply. I tried tf.random.set_seed(1) but it didn't work. I use version  
2.0.0b1 with keras, eager mode, GradientTape and GPU. I just tried the same code on the latest tf
version 2.0.0rc and got the same result where the memory keeps increasing to 10G after 5 epochs. My training data is dataframe and I passed it to each batch by generator. I print the dataframe.info and it is just 200K size. The batch size is 16 and I think it is not big. The model is seq2seq with LSTM. The memory is not released but I don't how to resolve it.
